### PR TITLE
Make SATNUM regions in FlowNet based on region parameter from simulation model

### LINKED ISSUE #392: Add option to select different region for relative permeability parameters
You can either have a global, regional or individual relative permeability scheme. The regional scheme leads to one relative permeability curve per `SATNUM` of the original model. However, in the Norne case there is only one `SATNUM` (endpoint scaling is used). It would be useful to allow the user to use a different region to create relative permeability curves for (for example `EQLNUM`)

### PR DESCRIPTION
**Pull request by olelod**
When setting the scheme to `regions_from_sim` the SATNUM regions in the input simulation model were used to generate the SATNUM regions for the FlowNet model. This PR makes this more flexible by allowing any region parameter from the INIT file to be used. 

### Contributor checklist

- [x] :tada: This PR closes #392.
- [x] :scroll: I have broken down my PR into the following tasks:
   - [x] Add region_parameter_from_sim_model to config_parser.py
   - [x] Add checks to config_parser.py
   - [x] Change the generation of SATNUM regions for FLOWNET to use the `region_parameter_from_sim_model` instead of the previously hard coded SATNUM
- [ ] :robot: I have added tests, or extended existing tests, to cover any new features or bugs fixed in this PR.
- [x] :book: I have considered adding a new entry in `CHANGELOG.md`.
- [ ] :books: I have considered updating the documentation.

### CODE REVIEW

File: src/flownet/config_parser/_config_parser.py [hunk 1]
```diff
@@ -1733,14 +1739,21 @@ def parse_config(
...
+            raise ValueError("Input simulation case is not defined. ")
         field_data = FlowData(config.flownet.data_source.simulation.input_case)
+        if config.model_parameters.relative_permeability.scheme == "regions_from_sim":
+            if (
+                config.model_parameters.relative_permeability.region_parameter_from_sim_model
+                not in field_data.init_headers
+            ):
+                raise ValueError(
+                    f"REGION parameter {config.model_parameters.relative_permeability.region_parameter_from_sim_model} "
+                    f"not found in input simulation model."
```
[the remark this answers is not in the document]
Comment (wouterjdb) [reply src/flownet/config_parser/_config_parser.py:1755]: Seems like it went trough... ok, that's a new thing in pylint?

**Comment by olelod** [comment]
Currently the implementation is such that you define a single prior which is used for all regions. I have not implemented a possibility for different priors in different regions (such as the one for equilibrium regions). I am open for doing that, but maybe make it an issue for further work?

**Comment by wouterjdb** [comment]
> Currently the implementation is such that you define a single prior which is used for all regions. I have not implemented a possibility for different priors in different regions (such as the one for equilibrium regions). I am open for doing that, but maybe make it an issue for further work?

That sounds like a good idea. Just make it an issue, and leave it on the backlog for the time-being.

File: src/flownet/config_parser/_config_parser.py [hunk 2]
```diff
@@ -1734,6 +1734,26 @@ def parse_config(
             "interpolation option for relative permeability."
         )
 
+    if (
+        config.model_parameters.equil.scheme != "regions_from_sim"
+        and config.model_parameters.equil.scheme != "individual"
+        and config.model_parameters.equil.scheme != "global"
```
Comment (wouterjdb) [inline src/flownet/config_parser/_config_parser.py:1741]: Shorter, and more readable?

```suggestion
        config.model_parameters.equil.scheme not in ["regions_from_sim", "individual", "global"]:
```

File: src/flownet/config_parser/_config_parser.py [hunk 3]
```diff
@@ -1734,6 +1734,26 @@ def parse_config(
...
+    ):
+        raise ValueError(
+            f"The equil scheme "
+            f"'{config.model_parameters.equil.scheme}' is not valid.\n"
+            f"Valid options are 'global', 'regions_from_sim' or 'individual'."
+        )
+    if (
+        config.model_parameters.relative_permeability.scheme != "regions_from_sim"
+        and config.model_parameters.relative_permeability.scheme != "individual"
+        and config.model_parameters.relative_permeability.scheme != "global"
```
Comment (wouterjdb) [inline src/flownet/config_parser/_config_parser.py:1751]: Same here.

File: src/flownet/config_parser/_config_parser.py [hunk 4]
```diff
@@ -1734,6 +1734,26 @@ def parse_config(
...
 
+    if (
+        config.model_parameters.equil.scheme != "regions_from_sim"
+        and config.model_parameters.equil.scheme != "individual"
+        and config.model_parameters.equil.scheme != "global"
+    ):
+        raise ValueError(
+            f"The equil scheme "
+            f"'{config.model_parameters.equil.scheme}' is not valid.\n"
+            f"Valid options are 'global', 'regions_from_sim' or 'individual'."
```
Comment (wouterjdb) [inline src/flownet/config_parser/_config_parser.py:1746]: In essence you _could_ use that list also here to generate the `ValueError` text. Not needed, just a thought.

File: src/flownet/config_parser/_config_parser.py [hunk 5]
```diff
@@ -1734,6 +1734,26 @@ def parse_config(
...
+        )
+    if (
+        config.model_parameters.relative_permeability.scheme != "regions_from_sim"
+        and config.model_parameters.relative_permeability.scheme != "individual"
+        and config.model_parameters.relative_permeability.scheme != "global"
+    ):
+        raise ValueError(
+            f"The relative permeability scheme "
+            f"'{config.model_parameters.relative_permeability.scheme}' is not valid.\n"
+            f"Valid options are 'global', 'regions_from_sim' or 'individual'."
```
Comment (wouterjdb) [inline src/flownet/config_parser/_config_parser.py:1756]: Same here.

File: src/flownet/config_parser/_config_parser.py [hunk 6]
```diff
@@ -1745,16 +1765,40 @@ def parse_config(
...
+            )
+        _check_if_all_region_priors_defined(
+            config.model_parameters.relative_permeability,
+            unique_satnum_regions,
+            "SATNUM",
+        )
+    else:
+        if config.model_parameters.relative_permeability.regions[0].id is not None:
+            raise ValueError(
+                "Id for firstrelative permeability region parameter should not be set, or set to 'None'\n"
```
Comment (wouterjdb) [inline src/flownet/config_parser/_config_parser.py:1788]:
```suggestion
                "The region number for the first relative permeability region parameter should not be set, or set to 'None'\n"
```

File: src/flownet/config_parser/_config_parser.py [hunk 7]
```diff
@@ -1745,16 +1765,40 @@ def parse_config(
...
+
+    if config.model_parameters.equil.scheme == "regions_from_sim":
+        unique_eqlnum_regions = field_data.get_unique_regions("EQLNUM")
+        _check_if_all_region_priors_defined(
+            config.model_parameters.equil, unique_eqlnum_regions, "EQLNUM"
+        )
+    else:
+        if config.model_parameters.equil.regions[0].id is not None:
+            raise ValueError(
+                "Id for first equilibrium region parameter should not be set, or set to 'None'\n"
```
Comment (wouterjdb) [inline src/flownet/config_parser/_config_parser.py:1800]:
```suggestion
                "The region number for the first equilibrium region parameter should not be set, or set to 'None'\n"
```

File: src/flownet/config_parser/_config_parser.py [hunk 8]
```diff
@@ -2333,3 +2317,45 @@ def _check_defined(path_in_config_dict: dict, parameter: str):
...
+):
+    """
+
+    Args:
+        path_in_config_dict: a location in the config schema dictionary
+        unique_regions: a list of the unique region numers in the input simulation model
+        parameter_name: the name of the output region parameter
+
+    Returns:
+        Nothing, raises ValueErrors if something is wrong
```
Comment (wouterjdb) [inline src/flownet/config_parser/_config_parser.py:2334]:
```suggestion
        Nothing

    Raises:
        ValueError: if something is wrong.
```
"if something is wrong" _could_ be slightly more specific 😃

File: src/flownet/config_parser/_config_parser.py (hunk 2 above)
Comment (olelod) [reply src/flownet/config_parser/_config_parser.py]: Surely!

File: src/flownet/config_parser/_config_parser.py (hunk 8 above)
Comment (olelod) [reply src/flownet/config_parser/_config_parser.py]: Yes, definitely. Changed the docstring in some other functions as well. Some can potentially raise many different ValueErrors. Not entirely clear to me how that should be written in the docstring, but an attempt here, at least.

File: src/flownet/config_parser/_config_parser.py [hunk 9]
```diff
@@ -1735,26 +1736,15 @@ def parse_config(
...
-        )
+    region_parameters = ["equil", "relative_permeability"]
+    for reg_param in region_parameters:
+        scheme = getattr(getattr(config.model_parameters, reg_param), "scheme")
+        if scheme not in available_region_schemes:
+            raise ValueError(
+                f"The {reg_param} scheme "
+                f"'{scheme}' is not valid.\n"
+                f"Valid options are {available_region_schemes}"
+            )
```
Comment (wouterjdb) [inline src/flownet/config_parser/_config_parser.py:1747]: Nice 👍

Comment (wouterjdb) [inline src/flownet/config_parser/_config_parser.py:1747]: Nice 👍

**Review by wouterjdb — APPROVED** [review]
LGTM 🚀

**Pull request merged**